hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OzoneManagerRequestHandler.java

```diff
@@ -1436,19 +1436,26 @@ private CancelSnapshotDiffResponse cancelSnapshotDiff(
   }
 
   private ListSnapshotDiffJobResponse listSnapshotDiffJobs(
-      ListSnapshotDiffJobRequest listSnapshotDiffJobRequest)
-      throws IOException {
-    List<SnapshotDiffJob> snapshotDiffJobs =
-        impl.listSnapshotDiffJobs(
-            listSnapshotDiffJobRequest.getVolumeName(),
-            listSnapshotDiffJobRequest.getBucketName(),
-            listSnapshotDiffJobRequest.getJobStatus(),
-            listSnapshotDiffJobRequest.getListAll());
-    ListSnapshotDiffJobResponse.Builder builder =
-        ListSnapshotDiffJobResponse.newBuilder();
-    for (SnapshotDiffJob diffJob : snapshotDiffJobs) {
+      ListSnapshotDiffJobRequest listSnapshotDiffJobRequest
+  ) throws IOException {
+    org.apache.hadoop.ozone.snapshot.ListSnapshotDiffJobResponse response = impl.listSnapshotDiffJobs(
+        listSnapshotDiffJobRequest.getVolumeName(),
+        listSnapshotDiffJobRequest.getBucketName(),
+        listSnapshotDiffJobRequest.getJobStatus(),
+        listSnapshotDiffJobRequest.getListAll(),
+        listSnapshotDiffJobRequest.getPrevSnapshotDiffJob(),
+        listSnapshotDiffJobRequest.getMaxListResult());
```

Comment:
To make it backward compatible, we need to check:

```java
int maxListResult = listSnapshotDiffJobRequest.hasMaxListResult()?listSnapshotDiffJobRequest.getMaxListResult():Integer.MAX_VALUE;
org.apache.hadoop.ozone.snapshot.ListSnapshotDiffJobResponse response = impl.listSnapshotDiffJobs(
        listSnapshotDiffJobRequest.getVolumeName(),
        listSnapshotDiffJobRequest.getBucketName(),
        listSnapshotDiffJobRequest.getJobStatus(),
        listSnapshotDiffJobRequest.getListAll(),
        listSnapshotDiffJobRequest.getPrevSnapshotDiffJob(),
        maxListResult);
```
or -1 denoting unlimited listing (need to handle this case in code). This way old clients will retain old behaviour of no pagination else it will not list any element.

Reply:
I can make the change, but [this validation](https://github.com/apache/ozone/pull/8124/files#diff-857b24769ac9c1c67d015bdffc1227dc2b4751de3790b9fd89269a3c46a5218aR836-R837) will not work because the client can send `Integer.MAX_VALUE` or `-1` no matter what we use to indicate unlimited listing.
Again, this API is only for Ops purposes and will mostly run on the Ozone cluster, so backward compatibility should not be a big concern. It is not like listing snapshots or keys. Snapshot diff jobs are short-lived and cleaned up after a certain time.

Reply:
We did a similar thing for `listKeys`/`listStatus` API in https://github.com/apache/ozone/pull/7431/files#diff-4e1ce03634e167877ec3c4d055a744b753d087ef7d9a3ed842f9936ef519aa71R775.

We can't enforce server-side validation till we have https://github.com/apache/ozone/pull/7603 or similar validation on the client side.

Reply:
In the listKeys PR you shared the max will be based on default value so it will atleast list the default max (config) but here old client will list 0 entries if the client is old .

> but the client can avoid setting MaxListResult, which will result in unlimited listing

This is how it is supposed to work right. If client does not pass the maxList arg it should print unlimited or if unlimited listing is a concern, at least limit it to the config (MAX_LIST_RESULT_DEFAULT)

To state clearly , if maxList arg is not set it will list 0 entries right now? Is this behaviour correct?

Reply:
The behavior is that there is no `unlimited` or `zero` listing. It will either list user-given number (capped to maxPageSize) or [maxPageSize](https://github.com/apache/ozone/pull/8124/files#diff-857b24769ac9c1c67d015bdffc1227dc2b4751de3790b9fd89269a3c46a5218aR202).

Check [this](https://github.com/apache/ozone/pull/8124/files#diff-857b24769ac9c1c67d015bdffc1227dc2b4751de3790b9fd89269a3c46a5218aR836-R838). If client doesn't pass `MaxListResult`, the server will return [maxPageSize](https://github.com/apache/ozone/pull/8124/files#diff-857b24769ac9c1c67d015bdffc1227dc2b4751de3790b9fd89269a3c46a5218aR202).

Reply:
I made the change to explicitly return `maxPageSize` entries if `MaxListResult` is not passed.

https://github.com/apache/ozone/pull/8124/commits/20b60bad97b45ddef9f0f1422ec9494aa4681923

Please take a look again.